### ISSUE TITLE
Invalid byte sequence error returned by Postgres

### ISSUE THREAD

**Issue by Th3-M4jor** [elixir-ecto/ecto_sql#520]
### Elixir version

elixir 1.14.4-otp-25

### Database and Version

Postgres 12.14, reproduced on Postgres 14.5

### Ecto Versions

Ecto: 3.10.1

### Database Adapter and Versions (postgrex, myxql, etc)

postgrex 0.17.1

### Current behavior

Migration:
```elixir
defmodule Test.Repo.Migrations.TestBinaryId do
  use Ecto.Migration

  def change do
    create table(:test_binary_id, primary_key: false) do
      add :id, :uuid, primary_key: true
    end
  end
end
```

Schema:
```elixir
defmodule Test.Schema do
  use Ecto.Schema

  @primary_key {:id, Ecto.UUID, autogenerate: true}
  schema "test_binary_id" do
  end
end
```

Inside an IEx shell with the above running do the following:
```
> import Ecto.Query
> from(t in Test.Schema, where: t.id == "507952ff-6234-4b53-aa82-638e72050aa7") |> Test.Repo.one
```

The below error is raised and IEx restarts:
```
** (EXIT from #PID<0.436.0>) shell process exited with reason: an exception was raised:
    ** (ArgumentError) errors were found at the given arguments:

  * 2nd argument: not valid character data (an iodata term)

        (stdlib 4.3.1) io.erl:99: :io.put_chars(:standard_io, [[<<27, 91, 51, 49, 109, 42, 42, 32, 40, 80, 111, 115, 116, 103, 114, 101, 120, 46, 69, 114, 114, 111, 114, 41, 32, 69, 82, 82, 79, 82, 32, 50, 50, 48, 50, 49, 32, 40, 99, 104, 97, 114, 97, 99, 116, 101, 114, 95, ...>>], 10, "\e[31m    (ecto_sql 3.10.1) lib/ecto/adapters/sql.ex:913: Ecto.Adapters.SQL.raise_sql_call_error/1\n    (ecto_sql 3.10.1) lib/ecto/adapters/sql.ex:828: Ecto.Adapters.SQL.execute/6\n    (ecto 3.10.1) lib/ecto/repo/queryable.ex:229: Ecto.Repo.Queryable.execute/4\n    (ecto 3.10.1) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3\n    (ecto 3.10.1) lib/ecto/repo/queryable.ex:151: Ecto.Repo.Queryable.one/3\n    iex:2: (file)\n\e[0m"])
        (iex 1.14.4) lib/iex/evaluator.ex:295: IEx.Evaluator.parse_eval_inspect/3
        (iex 1.14.4) lib/iex/evaluator.ex:187: IEx.Evaluator.loop/1
        (iex 1.14.4) lib/iex/evaluator.ex:32: IEx.Evaluator.init/4
        (stdlib 4.3.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
```
The visible part of the character sequence is `"\e[31m** (Postgrex.Error) ERROR 22021 (character_"`

Based on some further testing, this only occurs if the UUID string is part of the Ecto query itself. If its bound to a variable in the query or you instead use Repo.get/get_by no error is raised.

### Expected behavior

No exception is raised, or if one is raised it should not also cause IEx to restart

**Comment by greg-rychlewski** [comment]
Thanks for the report. The database error is happening because Postgres doesn't know `"507952ff-6234-4b53-aa82-638e72050aa7"` is supposed to be a UUID. If you use interpolation or get/get_by then Ecto knows its associated with the field and infers the type. At the moment I'm not sure if this can be improved but I can have a look.

I'm not too sure why it is failing in the way that you showed. Maybe it is trying to print the UUID binary and it's not valid. But Jose will know better about that.

**Comment by josevalim** [comment]
It seems the issue is that the invalid character is in the error message itself? So we need to make sure we can print the error message. I would start by inspecting it to see if it reveals anything else.

**Comment by greg-rychlewski** [comment]
I believe the error message is ok and it's just getting cutoff in their text `114, 95, ..>>`.  When I played around with it I was getting the `character_not_in_repertoire` error which matches what happens in Postgres when it receives an invalid UTF-8 byte.

From my investigation what happens is this causes the UUID string `507952ff-6234-4b53-aa82-638e72050aa7` to get tagged with type `t.id`

```elixir
t.id == "507952ff-6234-4b53-aa82-638e72050aa7"
```

Then in the planner when it gets to this part it is dumped to its 16 byte representation:
https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/query/planner.ex#L1290

Then later on when the adapter builds the SQL it is putting that 16 byte representation into it and it may not be valid UTF-8.

It's a bit tricky to see if this can be fixed. I have a feeling making the planner cast instead of dump will break something somewhere.

**Comment by greg-rychlewski** [comment]
Or I should say the error message is file until it hits the query string. Because Postgrex will insert the SQL into the error message.

**Comment by josevalim** [comment]
@greg-rychlewski ah, I am almost sure the Postgres.Connection was supposed to detect a binary and encode it using a string with escape sequences. so the SQL was supposed to be correct?

**Comment by greg-rychlewski** [comment]
Oh yes it is but it's already dumped by that point. So for this:

```elixir
from(t in Test.Schema, where: t.id == "507952ff-6234-4b53-aa82-638e72050aa7")
```

It produces this (39 is the escape):

```elixir
[
  39,
  <<80, 121, 82, 255, 98, 52, 75, 83, 170, 130, 99, 142, 114, 5, 10, 167>>,
  39
]
```

And for this

```elixir
q = from(t in TestID, where: t.bid == "22222222-2222-2222-2222-222222222222")
```

It produces this

```elixir
[39, "\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"\"", 39]
```

So the first one is basically putting quotes around bytes that can't be interpreted as UTF-8

**Comment by greg-rychlewski** [comment]
And the reason it is dumped when it gets to the adapter is because this makes the builder wrap the value in `%Ecto.Tagged{}` to say that the literal must be the same type as `t.id`:

```elixir
t.id == "507952ff-6234-4b53-aa82-638e72050aa7"
```

Then the planner sees `%Ecto.Tagged{}` and tries to convert it to the right type by dumping.

**Comment by josevalim** [comment]
Yeah, the connection was supposed to see the Ecto.Tagged and write a binary postgrex string with the bytes. Something like "\x32\xf3" and so on.

**Comment by greg-rychlewski** [comment]
Oh I see the part you are talking about. The problem is it is not tagged anymore when it gets to the connection module. It is just 16 bytes.

**Comment by josevalim** [comment]
Ah, we had code that was meant to keep it tagged all the way down. Maybe it is no longer working as expected.

**Comment by greg-rychlewski** [comment]
I can try to fix this. It seems like the planner is purposefully untagging everything in the prewalker but probably I don’t understand well enough yet .

**Comment by greg-rychlewski** [comment]
Oh actually it is not untagging base types. And it is considering any type that defences a field as not base. Maybe then something like this?

1. Evaluate the field type before checking if it’s a base type
2. Have to do something with :uuid type so it is still tagged.  It sure if it’s ok to consider it a base type atm. Or if it should just be a special case that remains tagged

**Comment by josevalim** [comment]
I wonder if we should convert `Ecto.UUID` type to `:binary`. I am not sure why it is `:uuid`. Maybe back then we could pass `Ecto.UUID` as a type to migration, but I don't think that's possible anymore.

**Comment by greg-rychlewski** [comment]
I tested out what happens when converting `Ecto.UUID` to `:binary` for this particular issue. It solves the problem with being invalid UTF-8 but still doesn't let you do the comparison. You get this error:

```elixir
** (Postgrex.Error) ERROR 42883 (undefined_function) operator does not exist: uuid = bytea
```

It looks to me like the only way comparison with UUID literals can work is if the string representation is used. What do you think about keeping the literal the way the user supplied it if the underlying type is `:uuid`? Basically this would force the user to use the string representation otherwise Postgres will complain about invalid comparison.

**Comment by greg-rychlewski** [comment]
Actually that won't work for MySQL because it doesn't have UUID it uses `binary(16)`. 

I feel like these need to be tagged with UUID type for each adapter to have its own rules.

**Comment by josevalim** [comment]
I see. So I guess we need to keep it as UUID, make sure they arrive tagged in the connection, so they raise a nice error message?

**Comment by greg-rychlewski** [comment]
I think we can do something similar to what SQL Server is doing:

```elixir
defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
    case binary do
      <<_::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96>> ->
        {:ok, value} = Tds.Ecto.UUID.dump(binary)
        value

      any ->
        any
    end
  end
```

But I think we have to dump in the planner because if someone makes a custom UUID type we will lose that information by the time we get to the connection module. So something like this for Postgres/MyXQL:

Postgres:

defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
    hex = Base.encode16(binary, case: :lower)
   [?', hex, ?' | "::uuid"]
 end
```

MySQL:

```elixir
defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
  hex = Base.encode16(binary, case: :lower)
  [?x, ?', hex, ?']
end
```

So basically Postgres will take the 16 byte representation and case it to the text representation. And MySQL will encode it as a byte string so it can be printed properly.

WDYT?

**Comment by josevalim** [comment]
Yup, that looks great!

**Comment by josevalim** [comment]
If they don't want fiddling, they need to tag it as `:binary`, otherwise I think it is safe to assume it is what the database takes it for `:uuid`?

**Comment by greg-rychlewski** [comment]
That makes sense. Okay I will start on this. Thanks for talking it out!

**Comment by mjaric** [comment]
> I'm not sure how kosher this is because it's assuming the byte order stays the same between text/binary for any UUID module someone might make:

It is not the same, you have varchar, that is not same as its  underlying system type storage (binary). I think along some lines pg got sting code points storage and not dumped Ecto.UUID binary value.

This works:
```sql
select * from table1 where id = 'a3a4f12f-e83c-4e93-988b-6950b3f945b7';
```

while this will NOT
```sql
DO $$
DECLARE
    p1 CONSTANT varchar(36) := 'a3a4f12f-e83c-4e93-988b-6950b3f945b7';

    BEGIN
        select * from table1 where id = p1;
    end;

$$;
```

**Comment by mjaric** [comment]
> I think we can get the query to work as expected by doing something similar to what SQL Server:
> 
> ```elixir
> defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
>     case binary do
>       <<_::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96>> ->
>         {:ok, value} = Tds.Ecto.UUID.dump(binary)
>         value
> 
>       any ->
>         any
>     end
>   end
> ```

This was necessary to overcome problem with Microsoft only standards :) it is not accidentally named GUID. Some bytes are in little endian storage while some are bigendian


> Variant 2 (110x2, N = c..d, 3 bits) is characterized in the RFC as "reserved, Microsoft Corporation backward compatibility" and was used for early GUIDs on the [Microsoft Windows](https://en.wikipedia.org/wiki/Microsoft_Windows) platform. It differs from variant 1 only by the endianness in binary storage or transmission: variant-1 UUIDs use "network" (big-endian) byte order, while variant-2 GUIDs use "native" (little-endian) byte order for some subfields of the UUID.

So, it was necessary to do something with that bytes that comes from elixir to mssql.

**Comment by greg-rychlewski** [comment]
>It is not the same, you have varchar, that is not same as its  underlying system type storage (binary). I think along some lines pg got sting code points storage and not dumped Ecto.UUID binary value.

Do you see any issue with this for Postgres:

1. The 16 byte representation is sent to the adapter
2. The adapter encodes it in base 16 to get its text representation in hexidecimal.
3. The adapter casts (2) to uuid `hex::uuid` and sents it to Postgres

**Comment by mjaric** [comment]
I would say 2. It should decode it in base 16.

And here is trace that proves it

```
(<0.278.0>) returned from 'Elixir.Ecto.Adapters.Postgres.Connection':all/1 -> [
      [],
      [<<"SELECT">>, [], 32,
        [[], [[112| <<"0">>], 46,34, <<"id">>, 34], <<", ">>], [112| <<"0">>], 46, 34,
        <<"name">>, 34], [<<" FROM ">>, [34,  <<"posts">>, 34], <<" AS ">>, 112| <<"0">>],
        [], [<<" WHERE ">>, 40, [[[112| <<"0">>], 46, 34, <<"id">>, 34], <<" = ">>, 39,
        <<51,209, 97,127, 66,254, 75,112, 130,216, 8,98,87, 109,89, 228>>,
        39], 41], [], [], [], [], [], [], []] 
```

` <<51,209, 97,127, 66,254, 75,112, 130,216, 8,98,87, 109,89, 228>>` is dumped value, and it is for postgres nonsense in query that it can't understand. It should be either string representation or parameter. Query must be valid string! Parameters are story for it self.

**Comment by mjaric** [comment]
More over, if you attempt to run logged query from some client
SELECT p0."id", p0."name" FROM "posts" AS p0 WHERE p0."id" = '3�a^B�Kp��^HbWmY�'

You will get
[2023-05-23 01:25:07] [22P02] ERROR: invalid input syntax for type uuid: "3�a^B�Kp��^HbWmY�"
[2023-05-23 01:25:07] Position: 62

**Comment by greg-rychlewski** [comment]
We are thinking not of sending just the bytes but tagged like this:

```elixir
%Ecto.Tagged{value: <<51,209, 97,127, 66,254, 75,112, 130,216, 8,98,87, 109,89, 228>>, type: :uuid}
```

Then when it gets to the Postgres connection module it will transform it to the string representation:

```elixir
defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
   hex = Base.encode16(binary, case: :lower)
   [?', hex, ?' | "::uuid"]
 end
```

So the query text would end up looking something like this:

```sql
SELECT p0."id", p0."name" FROM "posts" AS p0 WHERE p0."id" = '33d1617f42fe4b7082d80862576d59e4'::uuid
```

For MySQL it doesn't have a UUID type so it has to keep it as the bytes and would look like this:

```elixir
defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
   hex = Base.encode16(binary, case: :lower)
   [?x, ?', hex, ?']
 end
```

```sql
SELECT p0."id", p0."name" FROM "posts" AS p0 WHERE p0."id" = x'33d1617f42fe4b7082d80862576d59e4'
```

**Comment by mjaric** [comment]
If the value should be interpolated within query, then it should not be dumped. It should be a string, as command is, so it doesn't brake encoding of query command. Each db backend has own way to implicitly cast string from query while parsing query. This is common for all supported db backends in this repo. Exception is how one can specify explicit casting (fragments I guess), and it is bad to assume what user wants, we only have to sanitize strings parts that represent the value in command text. This does not applies to parametrized queries, since params have well defined network binary storage format and that yields best performance always if used.

So this
> ```elixir
> %Ecto.Tagged{value: <<51,209, 97,127, 66,254, 75,112, 130,216, 8,98,87, 109,89, 228>>, type: :uuid}
>```

Is both, inefficient and unnecessary since you are casting in elixir and in db backend while you could do only on backend if you passed it as string. Still, I understand the need to sanitize query, but if it is safe string (embedded in query command during compile time), then why bothering dumping it to binary?

Further more, lets consider date strings

```
from(t in Pgexp.Model.Post, where: t.created_at == "2000-01-01T23:00:07z") |> Pgexp.Repo.one
```
This doesn't work during runtime
```
** (Ecto.QueryError) lib/pgexp.ex:20: value `"2000-01-01T23:00:07z"` cannot be dumped to type :naive_datetime. Or the value is incompatible or it must be interpolated (using ^) so it may be cast accordingly in query:
```

Why? If we want to support query command interpolation, this is same as `select * from posts where created_at >  '2000-01-01T23:00:07z'`. Right?

So I don't think this should take the ecto value dump route, it is different problem, should have own feature. It can be new callback on Ecto.UUID module.

**Comment by josevalim** [comment]
> ** (Ecto.QueryError) lib/pgexp.ex:20: value `"2000-01-01T23:00:07z"` cannot be dumped to type :naive_datetime. Or the value is incompatible or it must be interpolated (using ^) so it may be cast accordingly in query:

The issue is that "Ecto" doesn't know, it asks the type, and the type chooses to dump or not. Unfortunately Ecto.UUID can dump (for reasons I can't recall) but I don't think we can change it without breaking code, so we either need to make it work or raise a better error message.

**Comment by greg-rychlewski** [comment]
I also don't think passing through a string will be enough for MySQL. It is expecting binary so the adapter would have to convert it to a byte string.

**Comment by mjaric** [comment]
So question is how to distinguish if parameter is literal or not and apply cast for non literals or something else for literals. :)

**Comment by mjaric** [comment]
This works for postgres adapter
```
from(t in Post, where: t.id == fragment("?", "33d1617f-42fe-4b70-82d8-0862576d59e4")) |> Repo.one
```

```
SELECT p0."id", p0."name", p0."created_at" FROM "posts" AS p0 WHERE (p0."id" = '33d1617f-42fe-4b70-82d8-0862576d59e4') []
%Pgexp.Model.Post{
  __meta__: #Ecto.Schema.Metadata<:loaded, "posts">,
  id: "33d1617f-42fe-4b70-82d8-0862576d59e4", 
  name: "test post",
  created_at: ~N[2023-05-23 07:47:54]
}
```

**Comment by greg-rychlewski** [comment]
From what it looks like to me, the literals are tagged with their expected type so that eventually they can be converted to a type that can be put in to a query. 

If they are a base type they remain tagged until the adapter and the adapter tries to cast it to the proper db type:

```elixir
defp expr(%Ecto.Query.Tagged{value: other, type: type}, sources, query) do
  [maybe_paren(other, sources, query), ?:, ?: | tagged_to_db(type)]
end
```

If they are not a base type then Ecto planner relies on the dumping from the adapter to convert it:

```elixir
defp prewalk(%Ecto.Query.Tagged{value: v, type: type} = tagged, kind, query, expr, acc, adapter) do
  if Ecto.Type.base?(type) do
    {tagged, acc}
  else
    {dump_param(kind, query, expr, v, type, adapter), acc}
    end
end
```

So to me it looks like there is a lot of attempts to convert the literals already and it's just what is the right way to do it for `:uuid`

**Comment by mjaric** [comment]
For mysql, all this works
```
select * from posts where id = `X'33d1617f42fe4b7082d80862576d59e4'`
select * from posts where id = 0x33d1617f42fe4b7082d80862576d59e4
select * from posts where id = '33d1617f-42fe-4b70-82d8-0862576d59e4'
```

pretty same like in MSSQL

**Comment by greg-rychlewski** [comment]
Yeah the last one will not return the correct results in MySQL. The first two are what happens now but I don't believe the user can pass that as a string I think we have to convert a string to those forms in the adapter.

**Comment by greg-rychlewski** [comment]
I think even if we could find some string representation that fits all adapters (I don't think we can but hypothetically), it looks like it would go against the rest of Ecto.

It seems on purpose that literals being compared with fields will be dumped to see if they are valid: https://github.com/elixir-ecto/ecto/blob/master/test/ecto/query/planner_test.exs#L1108

So let's say we will not refactor the above, then the UUID has to be dumped in the planner. And if it's dumped then the only way to set it correctly in the query string is to tag it.

**Comment by greg-rychlewski** [comment]
@josevalim Sorry for the ping :D just wanted to see where you stood on this.

I think based on the conversation so far the issue comes down to the fact that the planner is dumping any literals that are being compared to fields here:

```elixir
defp prewalk(%Ecto.Query.Tagged{value: v, type: type} = tagged, kind, query, expr, acc, adapter) do
  if Ecto.Type.base?(type) do
    {tagged, acc}
  else
    {dump_param(kind, query, expr, v, type, adapter), acc}
  end
end
```

I can only see 2 options personally:

1. What we already said, to tag the dumped UUIDs and have the adapters handle them properly.
2. Instead of dumping creating a new callback to transform literals for query strings so they can be passed through. Adapters like postgres would keep them strings. Adapters like MySQL would have to convert it to a printable binary. 

For 2 it seems like a lot of complications to me personally with limited use case so 1 is looking better.

**Comment by josevalim** [comment]
No problem. I prefer 1. If we find the back and forth expensive, the adapter may choose to raise and tell people to interpolate the param instead.

**Comment by josevalim** [comment]
Closing in favor of https://github.com/elixir-ecto/ecto/pull/4290.
